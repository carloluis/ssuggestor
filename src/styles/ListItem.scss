@import './_palette.scss';

:global .ss-over-item {
	background: $pale-grey;
}
