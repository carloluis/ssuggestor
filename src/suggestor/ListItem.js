import React, { PureComponent } from 'react';
import MarkItem from './MarkItem';

class ListItem extends PureComponent {
	render() {
		let { item, onItemClick, onItemMouseEnter, index, overItem, search } = this.props;
		return (
			<li value={item} onClick={() => onItemClick(item)} onMouseEnter={() => onItemMouseEnter(index, item)} 
				style={{ backgroundColor: overItem && '#f5f5f5' }}>
				<MarkItem {...{ item, search }} />
			</li>
		);
	}
}
ListItem.propTypes = {
	item: React.PropTypes.string.isRequired,
	index: React.PropTypes.number.isRequired,
	overItem: React.PropTypes.bool.isRequired,
	onItemClick: React.PropTypes.func.isRequired,
	onItemMouseEnter: React.PropTypes.func.isRequired,
	search: React.PropTypes.string.isRequired
};

export default ListItem;