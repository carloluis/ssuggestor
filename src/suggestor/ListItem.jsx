import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import autoBind from '../utils/auto-bind';
import MarkItem from './MarkItem';

class ListItem extends PureComponent {
	constructor(props) {
		super(props);
		autoBind(this);
	}

	handleClick(event) {
		event.stopPropagation();
		const { item, onItemClick } = this.props;
		onItemClick(item);
	}

	handleMouseEnter(event) {
		event.stopPropagation();
		const { index, onItemMouseEnter } = this.props;
		onItemMouseEnter(index);
	}

	render() {
		const {
			item,
			overItem,
			search,
			theme: { item: li, activeItem: liActive }
		} = this.props;
		const classes = [li, overItem && liActive].filter(classname => !!classname).join(' ');

		return (
			<li className={classes} onClick={this.handleClick} onMouseEnter={this.handleMouseEnter}>
				<MarkItem {...{ item, search }} />
			</li>
		);
	}
}

ListItem.propTypes = {
	item: PropTypes.object.isRequired,
	index: PropTypes.number.isRequired,
	overItem: PropTypes.bool.isRequired,
	onItemClick: PropTypes.func.isRequired,
	onItemMouseEnter: PropTypes.func.isRequired,
	search: PropTypes.string.isRequired,
	theme: PropTypes.shape({
		item: PropTypes.string,
		activeItem: PropTypes.string
	})
};

export default ListItem;
