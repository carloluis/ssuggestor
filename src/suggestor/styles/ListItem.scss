:global .ss-over-item {
	background: #eee;
}
