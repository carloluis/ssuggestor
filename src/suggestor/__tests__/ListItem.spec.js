import React from 'react';
import ListItem from '../ListItem';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

const PROPS = {
	item: { word: 'fructuary', index: -1 },
	index: 0,
	overItem: false,
	onItemClick: jest.fn(),
	onItemMouseEnter: jest.fn(),
	search: 'ss'
};

describe('<ListItem/>', () => {
	it('snapshot', () => {
		const tree = renderer.create(<ListItem {...PROPS} />).toJSON();
		expect(tree).toMatchSnapshot();
	});

	it('hover snapshot', () => {
		const tree = renderer.create(<ListItem {...PROPS} overItem />).toJSON();
		expect(tree).toMatchSnapshot();
	});
});

describe('ListItem component - item click', () => {
	const component = shallow(<ListItem {...PROPS} />);
	const event = {
		stopPropagation: jest.fn()
	};

	beforeEach(() => {
		event.stopPropagation.mockReset();
		PROPS.onItemClick.mockReset();
	});

	it('should call handleClick', () => {
		const handleClickSpy = jest.spyOn(component.instance(), 'handleClick');
		component.instance().forceUpdate();
		component.update();

		component.find('li').simulate('click', event);

		expect(handleClickSpy).toBeCalled();
		expect(PROPS.onItemClick).toBeCalled();
	});

	it('should stop event propagation', () => {
		component.find('li').simulate('click', event);

		expect(event.stopPropagation).toHaveBeenCalled();
	});

	it('should call props.onItemClick', () => {
		component.find('li').simulate('click', event);

		expect(PROPS.onItemClick).toBeCalledWith(PROPS.item);
	});
});

describe('ListItem component - mouse enter', () => {
	const component = shallow(<ListItem {...PROPS} />);
	const event = {
		stopPropagation: jest.fn()
	};

	beforeEach(() => {
		event.stopPropagation.mockReset();
		PROPS.onItemMouseEnter.mockReset();
	});

	it('should call handleMouseEnter', () => {
		const handleMouseEnterSpy = jest.spyOn(component.instance(), 'handleMouseEnter');
		component.instance().forceUpdate();
		component.update();

		component.find('li').simulate('mouseEnter', event);

		expect(handleMouseEnterSpy).toBeCalled();
	});

	it('should stop event propagation', () => {
		component.find('li').simulate('mouseEnter', event);

		expect(event.stopPropagation).toHaveBeenCalled();
	});

	it('should call props.onItemClick', () => {
		component.find('li').simulate('mouseEnter', event);

		expect(PROPS.onItemMouseEnter).toBeCalledWith(PROPS.index);
	});
});
