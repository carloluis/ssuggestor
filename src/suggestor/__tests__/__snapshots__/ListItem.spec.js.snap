// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListItem/> snapshot - hover 1`] = `
<li
  className="ss-over-item"
  onClick={[Function]}
  onMouseEnter={[Function]}
>
  <a>
    fructuary
  </a>
</li>
`;

exports[`<ListItem/> snapshot 1`] = `
<li
  className=""
  onClick={[Function]}
  onMouseEnter={[Function]}
>
  <a>
    fructuary
  </a>
</li>
`;
